crates/services/p2p/src/discovery/dnsaddr_resolution.rs

```diff
@@ -0,0 +1,101 @@
+pub trait DnsLookup {
+    async fn lookup_dnsaddr(&self, addr: &str) -> anyhow::Result<Vec<Multiaddr>>;
+    async fn new() -> anyhow::Result<Box<Self>>;
+}
+
+pub struct DnsResolver {
+    resolver: TokioAsyncResolver,
+}
+
+#[async_trait::async_trait]
+impl DnsLookup for DnsResolver {
+    async fn lookup_dnsaddr(&self, addr: &str) -> anyhow::Result<Vec<Multiaddr>> {
+        self.resolve_recursive(addr, 0).await
+    }
+
+    async fn new() -> anyhow::Result<Box<Self>> {
+        let resolver = TokioAsyncResolver::tokio_from_system_conf()?;
+        Ok(Box::new(Self { resolver }))
+    }
+}
+
+impl DnsResolver {
+    /// Internal method to handle recursive DNS lookups.
+    fn resolve_recursive<'a>(
+        &'a self,
+        addr: &'a str,
+        depth: usize,
+    ) -> Pin<
+        Box<dyn std::future::Future<Output = anyhow::Result<Vec<Multiaddr>>> + Send + 'a>,
+    > {
+        Box::pin(async move {
+            if depth >= MAX_DNS_LOOKUPS {
+                return Err(anyhow!("Maximum DNS lookup depth exceeded"));
+            }
+
+            let mut multiaddrs = vec![];
+            let dns_lookup_url = format!("{}{}", DNSADDR_PREFIX, addr);
+            let txt_records = self.resolver.txt_lookup(dns_lookup_url).await?;
+
+            for record in txt_records {
+                let parsed = record.to_string();
+                if !parsed.starts_with("dnsaddr") {
+                    continue;
+                }
+
+                let dnsaddr_value = parsed
+                    .split("dnsaddr=")
+                    .nth(1)
+                    .ok_or_else(|| anyhow!("Invalid DNS address: {:?}", parsed))?;
+
+                // Check if the parsed value is a multiaddress or another dnsaddr.
+                if dnsaddr_value.starts_with("/dnsaddr") {
+                    let nested_dnsaddr = dnsaddr_value
+                        .split('/')
+                        .nth(2)
+                        .ok_or_else(|| anyhow!("Invalid nested dnsaddr"))?;
+                    // Recursively resolve the nested dnsaddr
+                    #[allow(clippy::arithmetic_side_effects)]
+                    let nested_addrs =
+                        self.resolve_recursive(nested_dnsaddr, depth + 1).await?;
+                    multiaddrs.extend(nested_addrs);
+                } else if let Ok(multiaddr) = dnsaddr_value.parse::<Multiaddr>() {
+                    multiaddrs.push(multiaddr);
+                }
+            }
+
+            Ok(multiaddrs)
+        })
+    }
+}
+
+#[allow(non_snake_case)]
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[tokio::test]
+    async fn dns_resolver__parses_all_multiaddresses_from_ipfs_bootstrap_nodes() {
+        // given
+        let resolver = DnsResolver::new().await.unwrap();
+        // when
+        let multiaddrs = resolver
+            .lookup_dnsaddr("bootstrap.libp2p.io")
+            .await
+            .unwrap();
+        // then
+        assert!(!multiaddrs.is_empty());
```

Comment:
Do we assume here that any machine running this test will always be able to look up the "bootstrap.libp2p.io" address in their environment?

For example, what happens if I'm flushing my local DNS cache and disconnect from the internet before running this test? As I read this, the test case would fail in that scenario.

I think we need to introduce a a port to manage the DNS lookup to make this logic testable.

Reply:
I think introduction of the port will make this tests useless, since it will just test how port itself works. Plus I'm not sure how hard is it I think the introduction of the port will make these tests useless since they will just test how the port itself works. Plus I'm not sure how hard it is to write your own DNS resolver.

I'm okay with the idea that this test fails without a connection or local DNS resolution.

Having a test with real DNS proves that it works in a real environment. The libp2p library tests it in the same way=)

Reply:
While I see your point, I have to respectfully disagree. I think the port will help us make the tests **even more useful**. Right now we can't assert much more than `!multiaddrs.is_empty()`, because there's no way for us to control which DNS records we get when we look up `bootstrap.libp2p.io`. With the port, we for example can test that our logic parses the TXT records correctly.

As for another example, there's currently no way to know from this test if this test is doing a recursive lookup or just a plain single lookup - and this can vary depending on how `libp2p.io` configures their records. With the trait we can write multiple test cases to test and make clear assertions about different scenarios.

I don't think it would be too hard to put a trait between our code and the `TokioAsyncResolver`, because as far as I can see, we're only calling `TokioAsyncResolver::txt_lookup` so we only need to mock one method for the tests.

Reply:
Okay, I will try to put that in another way=)

The main functionality that I would like to see tested is that we can resolve the real addresses from ipfs(or fuel testnet/mainnet). So, I still want to see the test that uses real `dnsaddr` and verifies that `TokioAsyncResolver` works as expected.

The functionality of `resolve_recursive` and how this function works are not so important to me. 
The approach with the port for internal resolver can help to test the behaviour of the `resolve_recursive` function in different use cases. And it looks like the right call to do it.

It is up to @rymn. I'm okay with doing that in a separate PR just dedicated to better test coverage of the `resolve_recursive` since the main feature is implemented and we have an integration test that covers this feature=)

Reply:
how about this? I have some dnsaddr's on my personal domain anyway, we can just reuse those with recursion and assert strictly about the multiaddrs this function spits out?

Reply:
addressed in [8c77c1d](https://github.com/FuelLabs/fuel-core/pull/2204/commits/8c77c1d418b19c39a0d7b25c5d7f5a5b8abc6f7c)

i won't touch the records, i promise 😂

Reply:
Hahah pinky promise :joy: Nice comment also, super helpful thank you! :pray: